README.md
## [Flask API Server - Mongo](https://github.com/app-generator/api-server-flask-mongo)

Flask Starter with JWT authentication, and **SQLite** persistance - Provided by **AppSeed** [App Generator](https://appseed.us/app-generator).
It has all the ready-to-use bare minimum essentials.

<br />

> Features:

- ✅ `Up-to-date dependencies` 
- ✅ [API Definition](https://docs.appseed.us/boilerplate-code/api-unified-definition) - the unified API structure implemented by this server
- ✅ Simple JWT API powered by `flask-restx` and `flask_jwt_extended`
- ✅ Authentication with JWT (JWT login, JWT logout, Register)
- ✅ **Persistance**:  `MongoDB`
- ✅ Unitary tests, Docker

<br />

> Can be used with other [React Starters](https://appseed.us/apps/react) for a complete **Full-Stack** experience:

| [React Node JS Berry](https://appseed.us/product/berry-dashboard/api-server-nodejs/react/) | [React Node Soft Dashboard](https://appseed.us/product/soft-ui-dashboard/api-server-nodejs/react/) | [React Node Horizon](https://appseed.us/product/horizon-ui/api-server-nodejs/) |
| --- | --- | --- |
| [![React Node JS Berry](https://user-images.githubusercontent.com/51070104/176936514-f1bccb21-bafe-4b43-9e4c-b6fe0ec9511d.png)](https://appseed.us/product/berry-dashboard/api-server-nodejs/react/) | [![React Node Soft Dashboard](https://user-images.githubusercontent.com/51070104/176936814-74386559-4e05-43d5-b9a4-8f70ce96a610.png)](https://appseed.us/product/soft-ui-dashboard/api-server-nodejs/react/) | [![React Node Horizon](https://user-images.githubusercontent.com/51070104/174428337-181e6dea-0ad9-4fe1-a35f-25e5fa656a9d.png)](https://appseed.us/product/horizon-ui/api-server-nodejs/)

<br />

![Flask API Server - Open-source Flask Starter provided by AppSeed.](https://user-images.githubusercontent.com/51070104/126349643-264d4cf4-6d0b-4c24-8185-adf69409fa4e.png)

<br />

## Table of Contents

1. [Getting Started](#getting-started)
2. [Project Structure](#project-structure)
3. [Modules](#modules)
4. [Testing](#testing)


## Getting Started

clone the project

```bash
$ git clone https://github.com/app-generator/api-server-flask-mongo.git
$ cd api-server-flask-mongo
```

create virtual environment using python3 and activate it (keep it outside our project directory)

```bash
$ python3 -m venv /path/to/your/virtual/environment
$ source <path/to/venv>/bin/activate
```

install dependencies in virtualenv

```bash
$ pip install -r requirements.txt
```

setup `flask` command for our app

```bash
$ export FLASK_APP=run.py
$ export FLASK_ENV=development
```

> Or for Windows-based systems

```powershell
$ (Windows CMD) set FLASK_APP=run.py
$ (Windows CMD) set FLASK_ENV=development
$
$ (Powershell) $env:FLASK_APP = ".\run.py"
$ (Powershell) $env:FLASK_ENV = "development"
```

start test APIs server at `localhost:5000`

```bash
$ python run.py
```
or 
```bash
$ flask run
```

use `flask-restx`' swagger dashboard to test APIs, or use `POSTMAN`


## Project Structure

```bash
api-server-flask/
├── api
│   ├── config.py
│   ├── __init__.py
│   ├── models.py
│   └── routes.py
├── Dockerfile
├── README.md
├── requirements.txt
├── run.py
└── tests.py
```

<br />

## Docker

Our docker image details are in `Dockerfile`

To build the docker image (replace app name between `< >` as per your requirment)

```bash
$ docker build -t <api-server-flask-app>:latest .
```

To run the docker container

```bash
$ docker run -d -p 5000:5000 <api-server-flask-app>
```

<br />

## Modules

This application uses the following modules

 - Flask==1.1.4
 - flask-restx==0.4.0
 - Flask-JWT-Extended
 - flask-mongoengine
 - pytest

## Testing

Run tests using `pytest tests.py`

<br />

---
[Flask API Server - Mongo](https://github.com/app-generator/api-server-flask-mongo) - provided by AppSeed [App Generator](https://appseed.us)


CHANGELOG.md
# Change Log

## [v1.0.3] 2022-12-01
### Changes

- Deps Update
- Fix runtime issues

## [v1.0.2] 2022-06-18
### Improvements

- Update README 
- Open-Source Version

## [v1.0.0-rc2] 2021-08-21

### Stable release - Mongo branch

- Stack: Flask / Flask-RestX / Mongo / Docker
- API:
  - Sign UP: `/api/users/register`
  - Sign IN: `/api/users/login`
  - Logout: `/api/users/logout`
  - Check Session: `/api/users/checkSession`
  - Edit User: `/api/users/edit`
- Data persistence
  - Mongo
- Virtualization
  - Docker


tests.py
# -*- encoding: utf-8 -*-

"""
Copyright (c) 2019 - present AppSeed.us
"""

import pytest
import json

from api import app, db
from api.models import Users


"""
   Sample test data
"""

DUMMY_USERNAME = "apple"
DUMMY_USERNAME_NEW = "new new apple"
DUMMY_EMAIL = "apple@apple.com"
DUMMY_PASS = "newpassword"


@pytest.fixture
def client():

    with app.test_client() as client:
        yield client


def test_user_signup(client):
    """
       Tests /users/register API
    """
    response = client.post(
        "api/users/register",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME,
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 200
    assert "The user was successfully registered" in data["msg"]


def test_user_signup_invalid_data(client):
    """
       Tests /users/register API: invalid data like email field empty
    """
    response = client.post(
        "api/users/register",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME,
                "email": "",
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "'' is too short" in data["msg"]


def test_user_login_correct(client):
    """
       Tests /users/signup API: Correct credentials
    """
    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 200
    assert data["token"] != ""


def test_user_login_error(client):
    """
       Tests /users/signup API: Wrong credentials
    """
    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_EMAIL
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "Wrong credentials." in data["msg"]


def test_user_edit_correct_token(client):
    """
       Tests /users/edit API: Correct token
       First get token by logging in, and then pass token to edit API
    """

    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())

    _DUMMY_TOKEN = data["token"]

    response = client.post(
        "api/users/edit",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME_NEW
            }
        ),
        content_type="application/json",
        headers={'authorization': _DUMMY_TOKEN})

    data = json.loads(response.data.decode())
    assert response.status_code == 200


def test_user_edit_wrong_token(client):
    """
       Tests /users/edit API: Wrong token
    """
    response = client.post(
        "api/users/edit",
        data=json.dumps(
            {
                "username": DUMMY_USERNAME_NEW
            }
        ),
        content_type="application/json",
        headers={'authorization': "fake-token"})

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "Token is invalid" in data["msg"]


def test_check_session_correct_token(client):
    """
       Tests /users/checkSession API: Correct token
       First get token by logging in, and then pass token to checkSession API
    """

    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())

    _DUMMY_TOKEN = data["token"]

    response = client.post(
        "api/users/checkSession",
        data=json.dumps(
            {}
        ),
        content_type="application/json",
        headers={'authorization': _DUMMY_TOKEN})

    data = json.loads(response.data.decode())
    assert response.status_code == 200


def test_check_session_wrong_token(client):
    """
       Tests /users/checkSession API: Wrong token
    """
    response = client.post(
        "api/users/checkSession",
        data=json.dumps(
            {}
        ),
        content_type="application/json",
        headers={'authorization': "fake-token"})

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "Token is invalid" in data["msg"]


def test_logout_correct_token(client):
    """
       Tests /users/logout API: Correct token
       First get token by logging in, and then pass token to logout API
    """

    response = client.post(
        "api/users/login",
        data=json.dumps(
            {
                "email": DUMMY_EMAIL,
                "password": DUMMY_PASS
            }
        ),
        content_type="application/json")

    data = json.loads(response.data.decode())

    _DUMMY_TOKEN = data["token"]

    response = client.post(
        "api/users/logout",
        data=json.dumps(
            {}
        ),
        content_type="application/json",
        headers={'authorization': _DUMMY_TOKEN})

    data = json.loads(response.data.decode())
    assert response.status_code == 200


def test_logout_wrong_token(client):
    """
       Tests /users/logout API: Wrong token
    """
    response = client.post(
        "api/users/logout",
        data=json.dumps(
            {}
        ),
        content_type="application/json",
        headers={'authorization': "fake-token"})

    data = json.loads(response.data.decode())
    assert response.status_code == 400
    assert "Token is invalid" in data["msg"]


def test_drop_db(client):
    """
       Drop test database at end
    """

    with app.app_context():
        user = Users.get_by_email(DUMMY_EMAIL)
        user.delete()


api/models.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

from datetime import datetime

from werkzeug.security import generate_password_hash, check_password_hash
from flask_mongoengine import MongoEngine

db = MongoEngine() 

class Users(db.Document):
    username = db.StringField(max_length=32, required=True)
    email = db.StringField(required=True)
    password = db.StringField()
    jwt_auth_active = db.BooleanField()
    date_joined = db.DateTimeField(default=datetime.utcnow)

    def __repr__(self):
        return f"User {self.username}"

    def set_password(self, password):
        self.password = generate_password_hash(password)

    def check_password(self, password):
        return check_password_hash(self.password, password)

    def update_email(self, new_email):
        self.email = new_email

    def update_username(self, new_username):
        self.username = new_username

    def check_jwt_auth_active(self):
        return self.jwt_auth_active

    def set_jwt_auth_active(self, set_status):
        self.jwt_auth_active = set_status

    @classmethod
    def get_by_id(cls, id):
        return cls.query.get_or_404(id)

    @classmethod
    def get_by_email(cls, email):
        return cls.objects(email=email).first()

    def toDICT(self):

        cls_dict = {}
        cls_dict['_id'] = str(self.id)
        cls_dict['username'] = self.username
        cls_dict['email'] = self.email

        return cls_dict

    def toJSON(self):

        return self.toDICT()


class JWTTokenBlocklist(db.Document):
    jwt_token = db.StringField(required=True)
    created_at = db.DateTimeField(required=True)

    def __repr__(self):
        return f"Expired Token: {self.jwt_token}"

    @classmethod
    def get_by_jwt_token(cls, jwt_token):
        return cls.objects(jwt_token=jwt_token).first()


api/__init__.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

import json

from flask import Flask
from flask_cors import CORS

from .routes import rest_api
from .models import db

app = Flask(__name__)

app.config.from_object('api.config.BaseConfig')

db.init_app(app)
rest_api.init_app(app)
CORS(app)

"""
   Custom responses
"""

@app.after_request
def after_request(response):
    """
       Sends back a custom error with {"success", "msg"} format
    """

    if int(response.status_code) >= 400:
        response_data = json.loads(response.get_data())
        if "errors" in response_data:
            response_data = {"success": False,
                             "msg": list(response_data["errors"].items())[0][1]}
            response.set_data(json.dumps(response_data))
        response.headers.add('Content-Type', 'application/json')
    return response


api/routes.py
# -*- encoding: utf-8 -*-
"""
Copyright (c) 2019 - present AppSeed.us
"""

from datetime import datetime, timezone, timedelta

from functools import wraps

from flask import request
from flask_restx import Api, Resource, fields

import jwt

from .models import db, Users, JWTTokenBlocklist
from .config import BaseConfig

rest_api = Api(version="1.0", title="Users API")


"""
    Flask-Restx models for api request and response data
"""

signup_model = rest_api.model('SignUpModel', {"username": fields.String(required=True, min_length=2, max_length=32),
                                              "email": fields.String(required=True, min_length=4, max_length=64),
                                              "password": fields.String(required=True, min_length=4, max_length=16)
                                              })

login_model = rest_api.model('LoginModel', {"email": fields.String(required=True, min_length=4, max_length=64),
                                            "password": fields.String(required=True, min_length=4, max_length=16)
                                            })

user_edit_model = rest_api.model('UserEditModel', {"userID": fields.String(required=True, min_length=1, max_length=32),
                                                   "username": fields.String(required=True, min_length=2, max_length=32),
                                                   "email": fields.String(required=True, min_length=4, max_length=64)
                                                   })

"""
   Helper function for JWT token required
"""

def token_required(f):

    @wraps(f)
    def decorator(*args, **kwargs):

        token = None

        if "authorization" in request.headers:
            token = request.headers["authorization"]

        if not token:
            return {"success": False, "msg": "Valid JWT token is missing"}, 400

        try:
            data = jwt.decode(token, BaseConfig.SECRET_KEY, algorithms=["HS256"])
            current_user = Users.get_by_email(data["email"])

            if not current_user:
                return {"success": False,
                        "msg": "Sorry. Wrong auth token. This user does not exist."}, 400

            token_expired = JWTTokenBlocklist.get_by_jwt_token(token)

            if token_expired is not None:
                return {"success": False, "msg": "Token revoked."}, 400

            if not current_user.check_jwt_auth_active():
                return {"success": False, "msg": "Token expired."}, 400

        except:
            return {"success": False, "msg": "Token is invalid"}, 400

        return f(current_user, *args, **kwargs)

    return decorator


"""
    Flask-Restx routes
"""


@rest_api.route('/api/users/register')
class Register(Resource):
    """
       Creates a new user by taking 'signup_model' input
    """

    @rest_api.expect(signup_model, validate=True)
    def post(self):

        req_data = request.get_json()

        _username = req_data.get("username")
        _email = req_data.get("email")
        _password = req_data.get("password")

        user_exists = Users.get_by_email(_email)
        if user_exists:
            return {"success": False,
                    "msg": "Email already taken"}, 400

        new_user = Users(username=_username, email=_email)

        new_user.set_password(_password)
        new_user.save()

        return {"success": True,
                "userID": str(new_user.id),
                "msg": "The user was successfully registered"}, 200


@rest_api.route('/api/users/login')
class Login(Resource):
    """
       Login user by taking 'login_model' input and return JWT token
    """

    @rest_api.expect(login_model, validate=True)
    def post(self):

        req_data = request.get_json()

        _email = req_data.get("email")
        _password = req_data.get("password")

        user_exists = Users.get_by_email(_email)

        if not user_exists:
            return {"success": False,
                    "msg": "This email does not exist."}, 400

        if not user_exists.check_password(_password):
            return {"success": False,
                    "msg": "Wrong credentials."}, 400

        # create access token uwing JWT
        token = jwt.encode({'email': _email, 'exp': datetime.utcnow() + timedelta(minutes=30)}, BaseConfig.SECRET_KEY)

        user_exists.set_jwt_auth_active(True)
        user_exists.save()

        return {"success": True,
                "token": token,
                "user": user_exists.toJSON()}, 200


@rest_api.route('/api/users/edit')
class EditUser(Resource):
    """
       Edits User's username or password or both using 'user_edit_model' input
    """

    @rest_api.expect(user_edit_model)
    @token_required
    def post(self, current_user):

        req_data = request.get_json()

        _new_username = req_data.get("username")
        _new_email = req_data.get("email")

        if _new_username:
            self.update_username(_new_username)

        if _new_email:
            self.update_email(_new_email)

        self.save()

        return {"success": True}, 200


@rest_api.route('/api/users/checkSession')
class checkSession(Resource):
    """
       Checks an existing JWT Token for validity
    """

    @token_required
    def post(self, current_user):

        if self:
            return {"success": True, "msg": "User is logged on with valid token"}, 200
        else:
            return {"success": False, "msg": "User is not logged on"}, 400


@rest_api.route('/api/users/logout')
class LogoutUser(Resource):
    """
       Logs out User using 'logout_model' input
    """

    @token_required
    def post(self, current_user):

        _jwt_token = request.headers["authorization"]

        jwt_block = JWTTokenBlocklist(jwt_token=_jwt_token, created_at=datetime.now(timezone.utc))
        jwt_block.save()

        self.set_jwt_auth_active(False)
        self.save()

        return {"success": True}, 200
